### PR TITLE
[Android] Restore end-to-end notifications on Mentra Live (OS-1821)

### PR DESCRIPTION
**Pull request by isaiahb**
Mentra Live has a speaker and no display, so the reference card the display path requests is invisible there and the wearer learns nothing. This reads the notification aloud instead, and fixes the priority signal that every consumer depends on.

## Reading notifications aloud

`speakPhoneNotification()` is deliberately conservative, because a notification written to be *seen* is hostile when read verbatim:

* Silent on glasses that **do** have a display — the card is better than speech.
* Requires an on-device TTS voice; no network round trip per notification.
* `buildSpokenNotification()` strips URLs and emoji, cuts on a word boundary, and drops summary rows.
* A burst is summarised rather than recited. Notifications do not arrive politely spaced — apps re-post to update a download or a music player, and granting notification access replays whatever is already in the shade. Anything landing inside a 10s quiet window is counted and summarised once ("4 more notifications") instead of read out.

## Priority was a hardcoded placeholder

`onNotificationPosted` reported a fixed value for every notification, so every consumer's priority handling was dead code — including the do-not-interrupt gate in `buildSpokenNotification()`.

The fix reads channel importance from the ranking rather than `Notification.priority`, which was deprecated at API 26 and reads `0` for apps that express intent through a channel — which would make those look ordinary. Only the service can see the ranking, so it is resolved there and passed down, mapped onto the scale consumers already understand (`importance - 3`, so LOW/MIN go negative), falling back to the deprecated field only when no ranking is available.

## Verified on device

Moto g play 2024 (Android 14), listener bound:

```
CrustNotificationListener: Received notification from com.android.shell
CrustNotificationListener: Processing buffered notification from Shell
MANTLE: phone_notification from Shell, priority=0
```

`0` is `IMPORTANCE_DEFAULT` (3) through the mapping — a real value resolved from the channel ranking, where the previous code reported a placeholder. Consistent across repeated posts.

**Not verified:** the negative case. `cmd notification post` always lands on the shell's default-importance channel, so a LOW-channel notification being dropped at the gate has not been observed on device. The arithmetic is trivial and the default case is confirmed live, but that half is reasoning rather than evidence.

`bun test` green, `tsc` clean.

## Note for reviewers

Getting a notification to reach this code at all requires the listener component to be enabled, which lives behind the Debug settings screen. On a device where it is off, `pm dump` shows:

```
disabledComponents:
  com.mentra.crust.services.NotificationListenerServiceImpl
```

and `adb shell pm enable` cannot fix it (`SecurityException: Shell cannot change component state`) — only the app can enable its own component. Worth knowing before concluding notifications are broken.

---

> [!NOTE]
> **Medium Risk**
> Touches notification listener, Mantle lifecycle/TTS playback, and a breaking type change for `priority` (string → number); behavior is gated by hardware capabilities but logout/cleanup ordering matters for audio state.
> 
> **Overview**
> **Mentra Live (speaker, no display)** now gets audible notification alerts instead of invisible reference cards. `MantleManager` speaks notifications only when capabilities lack a display but have a speaker; display glasses still use the card path.
> 
> **Native priority is fixed end-to-end.** `CrustModule` no longer emits a hardcoded `"normal"` string; it forwards a numeric Android priority. The listener resolves **channel importance** from `NotificationListenerService` ranking (mapped as `importance - 3`), with fallback to deprecated `Notification.priority`. `PhoneNotificationEvent.priority` is typed as `number`.
> 
> **Speech pipeline:** new `spokenNotification` helper sanitizes text (URLs, emoji, summaries), enforces a priority floor, and caps length. `MantleManager` adds TTS playback with a 10s quiet window, burst summarization ("N more notifications"), concurrency/cleanup guards, and explicit teardown on logout.
> 
> **Notify miniapp** no longer requires `DISPLAY` hardware (optional), so notification capture works on Live.
> 
> Tests cover spoken-notification shaping rules.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c91793a43e75174f3769361d24774b03a0ae6d89. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: mobile/src/services/MantleManager.ts [hunk 1]
```diff
@@ -88,6 +115,103 @@ class MantleManager {
...
+  /** Synthesize and play one line. Assumes the caller checked the quiet window. */
+  private async speakAloud(text: string): Promise<void> {
+    let audio: {audioUrl: string; cleanup: () => Promise<void>} | undefined
+    try {
+      if (!(await ttsModelManager.isModelAvailable())) {
+        console.warn("MantleManager: no on-device TTS voice available, notification not spoken")
+        return
+      }
+
+      this.speakingNotification = true
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply mobile/src/services/MantleManager.ts]: Valid, fixed in a8820519ba. Cursor found this independently too, which is a fair signal that it was the real defect in this PR.

Confirmed in the code: \`speakingNotification\` was set only after \`await ttsModelManager.isModelAvailable()\`, while the caller's quiet-window check reads that same flag. A burst therefore passed the check together, synthesized concurrently, and interrupted one another — and the first completion callback then cleared the shared flag while a later clip was still playing.

Now claimed synchronously at the top of \`speakAloud\`, before anything awaits. A check-and-set with no await between the two is atomic here, so that is the whole fix. The stuck guard is armed at the same point and released on every exit path, including the model-unavailable early return.

[the remark this answers is not in the document]
Comment (isaiahb) [reply mobile/src/services/MantleManager.ts]: Valid, fixed in a8820519ba — same defect Codex flagged, and you both landed on it independently.

Your second sentence is the part that made it worth fixing rather than tolerating: not just that two clips overlap, but that the interrupted completion callback clears the shared flag while the newer clip is still playing, so the gate is wrong for everything that follows too.

The slot is now claimed synchronously before the first await.

File: mobile/src/services/MantleManager.ts [hunk 2]
```diff
@@ -88,6 +115,103 @@ class MantleManager {
...
+    if (this.suppressedSummaryTimer !== null) return
+    this.suppressedSummaryTimer = BgTimer.setTimeout(() => {
+      this.suppressedSummaryTimer = null
+      const count = this.suppressedNotifications
+      this.suppressedNotifications = 0
+      if (count <= 0) return
+      if (this.speakingNotification) return
+      void this.speakAloud(`${count} more notification${count === 1 ? "" : "s"}.`)
+    }, SPOKEN_NOTIFICATION_GAP_MS)
+  }
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply mobile/src/services/MantleManager.ts:186]: Valid and the most consequential of the set. Fixed in a8820519ba.

The ordering was exactly as described — zero the count, then check \`speakingNotification\` and return — so the count was discarded rather than deferred. Worth adding why it was not an edge case: the timer is anchored to the first suppressed arrival, while the quiet window only starts when speech ENDS. Firing while speech is still in progress is therefore the normal path, not a rare one, so a burst landing mid-announcement — precisely what the summary exists for — reliably produced no summary at all.

Now it reschedules and keeps the count, and the reschedule condition includes the post-playback quiet window so it re-anchors to \`lastSpokenAt\` rather than to arrival.

File: mobile/src/services/MantleManager.ts [hunk 3]
```diff
@@ -88,6 +115,103 @@ class MantleManager {
...
+   * "Four more notifications" is useful; four notifications read in full is not.
+   */
+  private scheduleSuppressedSummary(): void {
+    if (this.suppressedSummaryTimer !== null) return
+    this.suppressedSummaryTimer = BgTimer.setTimeout(() => {
+      this.suppressedSummaryTimer = null
+      const count = this.suppressedNotifications
+      this.suppressedNotifications = 0
+      if (count <= 0) return
+      if (this.speakingNotification) return
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply mobile/src/services/MantleManager.ts]: Valid, fixed in a8820519ba, together with the count-dropping issue Cursor raised on the same function.

Both halves you describe are addressed: the timer no longer trusts its original anchor, and instead re-checks \`Date.now() - lastSpokenAt\` against the gap on each firing, rescheduling while speech is active or the quiet window has not elapsed. \`lastSpokenAt\` is updated from the playback completion callback, so this is effectively anchored to completion as you suggested. The count is retained across reschedules rather than reset.

File: mobile/src/services/MantleManager.ts [hunk 4]
```diff
@@ -78,6 +98,13 @@ class MantleManager {
   private subs: Array<any> = []
   private initialized: boolean = false
   private activePhoneNotificationId: string | null = null
+  /** A notification is being read aloud right now. */
+  private speakingNotification: boolean = false
+  /** When the last announcement finished, for the quiet window. */
+  private lastSpokenAt: number = 0
+  /** Notifications that arrived during the quiet window, summarised afterwards. */
+  private suppressedNotifications: number = 0
+  private suppressedSummaryTimer: ReturnType<typeof BgTimer.setTimeout> | null = null
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply mobile/src/services/MantleManager.ts:107]: Valid, fixed in a8820519ba.

\`cleanup()\` handled the calendar timer and subscriptions but nothing this feature added, so a queued summary could synthesize and speak after the subscriptions that produced it were gone, or carry its count and speaking flag into the next login in the same process — and that logout-then-login-without-a-restart path is real, it is the one exercised by #3633.

Now clears the summary timer, resets the counters, flag and \`lastSpokenAt\`, and calls \`audioPlaybackService.stopForApp(notifyPackageName)\`. You are right that \`engine.stop()\` does not cover playback, so that call is explicit.

File: mobile/src/services/MantleManager.ts [hunk 5]
```diff
@@ -88,6 +115,103 @@ class MantleManager {
...
+          // Start the quiet window when speech ENDS, not when it started, so a
+          // long announcement isn't immediately followed by another.
+          this.lastSpokenAt = Date.now()
+          void audio?.cleanup?.()
+        },
+      )
+    } catch (error) {
+      this.speakingNotification = false
+      void audio?.cleanup?.()
+      console.warn("MantleManager: failed to speak notification audio", error)
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply mobile/src/services/MantleManager.ts:265]: Valid, fixed in a8820519ba.

\`stuckGuard\` was cleared only from the playback completion callback, so a throw from \`synthesizeToFile\` left it armed while the catch reset the flag. It would then fire up to \`SPOKEN_NOTIFICATION_MAX_MS\` later and clear \`speakingNotification\` for an announcement that was by then legitimately playing — turning a one-off synthesis failure into an overlap much later.

The guard and the flag are now released together through a single \`release()\` used by the success path, the catch, and the model-unavailable early return.

**Comment by isaiahb** [comment]
All six bot findings were valid and are fixed in `a8820519ba`. Both bots independently flagged the same two core defects, which is a fair signal they were the real problems.

They cluster into one theme — the burst handling, which is the part of this feature that actually justifies its existence:

| Finding | Effect |
|---|---|
| Slot claimed after `await` (Codex P1 + Cursor High) | Burst passes the gate together, clips interrupt each other, first completion clears the flag for a clip still playing |
| Suppressed count dropped (Cursor High + Codex P2) | Burst landing mid-announcement produced **no** summary — the exact case the summary is for |
| Stuck guard leaks on synthesis failure (Cursor Medium) | One-off failure causes an overlap up to 30s later |
| `cleanup()` ignores this state (Codex P2) | Summary speaks after teardown, or count carries into the next login in the same process |

The count-dropping one is worth calling out: the timer is anchored to the first suppressed arrival while the quiet window only starts when speech *ends*, so firing early was the normal path rather than an edge case. That feature was effectively broken whenever it mattered most.

## Verification, stated precisely

These four are fixed **by inspection and typecheck, not on device**. Exercising the burst path needs speaker-only glasses (Mentra Live) and the phone under test has none paired, so I have not watched a real burst suppress and summarise correctly. The priority mapping this PR also carries **is** device-verified — see the description.

`bun test` 598 passing, `tsc` clean.

File: mobile/src/services/MantleManager.ts [hunk 6]
```diff
@@ -88,6 +115,132 @@ class MantleManager {
...
+          // long announcement isn't immediately followed by another.
+          this.lastSpokenAt = Date.now()
+          void audio?.cleanup?.()
+        },
+      )
+    } catch (error) {
+      release()
+      void audio?.cleanup?.()
+      console.warn("MantleManager: failed to speak notification audio", error)
+    }
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply mobile/src/services/MantleManager.ts:266]: Valid, fixed in c91793a43e.

The stamp before `synthesizeToFile` is deliberate — it gives the stuck guard a sane window — but you are right that `release()` left it behind on the failure paths, so a quiet window opened for speech that never happened and the next 10s of notifications were summarised for it.

Now captured as `previousSpokenAt` and rolled back on every path that does not actually announce: the catch, and the stale-generation bail-out added for the other finding. The model-unavailable early return already sits above the stamp, so it was never affected.

File: mobile/src/services/MantleManager.ts [hunk 7]
```diff
@@ -243,6 +396,19 @@ class MantleManager {
...
+    // speaking flag into the next login in the same process. engine.stop() does
+    // not stop playback, so the notify-owned audio has to be stopped explicitly.
+    if (this.suppressedSummaryTimer !== null) {
+      BgTimer.clearTimeout(this.suppressedSummaryTimer)
+      this.suppressedSummaryTimer = null
+    }
+    this.suppressedNotifications = 0
+    this.speakingNotification = false
+    this.lastSpokenAt = 0
+    void audioPlaybackService.stopForApp(notifyPackageName)
```
[the remark this answers is not in the document]
Comment (isaiahb) [reply mobile/src/services/MantleManager.ts]: Valid, and I confirmed the mechanism rather than taking it on faith. `stopForApp` → `interruptCurrentPlayback`, which is documented as "Interrupt current playback **and notify its callback**" — so my completion callback ran and re-stamped `lastSpokenAt` immediately after `cleanup()` zeroed it. Fixed in c91793a43e.

Playback is now stopped **before** the state reset, and awaited rather than fired and forgotten. Ordering was the whole bug.

Your second point — in-flight `synthesizeToFile` still reaching `play()` after cleanup — was the more interesting one, since no ordering change fixes it. Added a generation token bumped by `cleanup()`; `speakAloud` captures it on entry and checks it both before `play()` and inside the completion callback before stamping. Stale work releases its slot, rolls the timestamp back, cleans up its audio file, and returns without playing.

Worth noting this is the same failure family as the previous round: state that was previously self-correcting because something else overwrote it, and is not once the lifecycle is handled properly.

**Comment by aisraelov** [comment]
Superseded by #3648, which preserves this PR's TTS/priority/sanitization/burst work and integrates it with Notify-gated presentation plus the Android listener lifecycle/ANR fix. Closing this one so review can happen in one place. @isaiahb
